### PR TITLE
[SPARK-20231] [SQL] Refactor star schema code for the subsequent star join detection in CBO

### PR DESCRIPTION
**Pull request by ioana-delaney**
## What changes were proposed in this pull request?

This commit moves star schema code from ```join.scala``` to ```StarSchemaDetection.scala```. It also applies some minor fixes in ```StarJoinReorderSuite.scala```.

## How was this patch tested?
Run existing ```StarJoinReorderSuite.scala```.

### CODE REVIEW

**Comment by ioana-delaney** [comment]
@gatorsmile I did a small refactoring for star schema. Would you please review. Thank you.

**Comment by gatorsmile** [comment]
add to whitelist

**Comment by gatorsmile** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/StarSchemaDetection.scala [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+   *
+   * Next, it computes the set of dimension tables for the current fact table. A dimension table
+   * is assumed to be in a RI relationship with a fact table. To infer column uniqueness,
+   * the algorithm compares the number of distinct values with the total number of rows in the
+   * table. If their relative difference is within certain limits (i.e. ndvMaxError * 2, adjusted
+   * based on 1TB TPC-DS data), the column is assumed to be unique.
+   */
+  def findStarJoins(
+      input: Seq[LogicalPlan],
+      conditions: Seq[Expression]): Seq[LogicalPlan] = {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/StarSchemaDetection.scala]: To the other reviewers, this is to address the comment by @cloud-fan  in the previous PR. 
https://github.com/apache/spark/pull/15363#discussion_r106774155

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala [hunk 2]
```diff
@@ -20,339 +20,13 @@ package org.apache.spark.sql.catalyst.optimizer
...
-    } else {
-      val starPlan = starPlans.head
-      val (factTable, dimTables) = (starPlan.head, starPlan.tail)
-
-      // Only consider selective joins. This case is detected by observing local predicates
-      // on the dimension tables. In a star schema relationship, the join between the fact and the
-      // dimension table is a FK-PK join. Heuristically, a selective dimension may reduce
-      // the result of a join.
-      // Also, conservatively assume that a fact table is joined with more than one dimension.
-      if (dimTables.size >= 2 && isSelectiveStarJoin(dimTables, conditions)) {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala]: `dimTables.size >= 2` is removed in the new code? Could you please explain it?

Comment (ioana-delaney) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala]: @gatorsmile I moved the condition from ```reorderStarJoin``` to ```findStarJoin``` to be used by both ReorderJoin and CostBasedReorder.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/StarSchemaDetection.scala [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+              val eligibleDimPlans = allFactJoins.filter { dimTable =>
+                allFactJoinCond.exists {
+                  case cond @ Equality(lhs: AttributeReference, rhs: AttributeReference) =>
+                    val dimCol = if (dimTable.outputSet.contains(lhs)) lhs else rhs
+                    isUnique(dimCol, dimTable)
+                  case _ => false
+                }
+              }
+
+              if (eligibleDimPlans.isEmpty || eligibleDimPlans.size < 2) {
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/StarSchemaDetection.scala]: uh, you move it to here.

**Comment by gatorsmile** [comment]
LGTM. Look forward to your next PR for merging start join detection with CBO.

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by ioana-delaney** [comment]
@gatorsmile Thank you!

**Pull request closed without merging**